lib/state_change_syncer.go

```diff
@@ -0,0 +1,431 @@
+			stateChangeEntry.KeyBytes = connectedMempoolTx.KeyBytes
+			stateChangeEntry.UtxoOps = connectedMempoolTx.UtxoOps
+		} else {
+			return
+		}
+	}
+
+	// Set the encoder type.
+	stateChangeEntry.EncoderType = stateChangeEntry.Encoder.GetEncoderType()
+
+	// Encode the state change entry. They are encoded as a byte array, so the consumer can buffer just the bytes needed
+	// to decode this entry when reading from file.
+	entryBytes := EncodeToBytes(0, stateChangeEntry, false)
+	writeBytes := EncodeByteArray(entryBytes)
+
+	// All mempool transactions occur within the same thread, so they'll be on the uuid.Nil flush ID.
+	stateChangeSyncer.addTransactionToQueue(uuid.Nil, writeBytes)
+
+	if event.IsConnected {
+		// With mempool connects, instantly flush to db.
+		// As soon as this callback is called, the mempool transaction is considered "connected".
+		err := stateChangeSyncer.FlushTransactionsToFile(&DBFlushedEvent{FlushId: uuid.Nil, Succeeded: true})
+		if err != nil {
+			glog.Fatalf("StateChangeSyncer._handleMempoolTransaction: Error flushing mempool transaction to file: %v", err)
+		}
+	} else {
+		// After the disconnect is written to file, remove this mempool tx from our map.
+		delete(stateChangeSyncer.ConnectedMempoolTxMap, *event.TxHash)
+	}
+}
+
+// handleDbTransactionConnected is called when a badger db operation takes place.
+// This function checks to see if the operation effects a "core_state" index, and if so it encodes a StateChangeEntry
+// to be written to the state change file upon DB flush.
+// It also writes the offset of the entry in the file to a separate index file, such that a consumer can look up a
+// particular entry index in the state change file.
+func (stateChangeSyncer *StateChangeSyncer) _handleDbTransaction(event *DBTransactionEvent) {
+	stateChangeSyncer.StateSyncerMutex.Lock()
+	defer stateChangeSyncer.StateSyncerMutex.Unlock()
+
+	stateChangeEntry := event.StateChangeEntry
+
+	// Check to see if the index in question has a "core_state" annotation in its definition.
+	if !isCoreStateKey(stateChangeEntry.KeyBytes) {
+		return
+	}
+
+	// Get the relevant deso encoder for this keyBytes.
+	var encoderType EncoderType
+	if isEncoder, encoder := StateKeyToDeSoEncoder(stateChangeEntry.KeyBytes); isEncoder && encoder != nil {
+		encoderType = encoder.GetEncoderType()
+	} else {
+		// If the keyBytes is not an encoder, then we decode the entry from the key value.
+		keyEncoder, err := DecodeStateKey(stateChangeEntry.KeyBytes)
+		if err != nil {
+			glog.Fatalf("Server._handleDbTransaction: Error decoding state key: %v", err)
+		}
+		encoderType = keyEncoder.GetEncoderType()
+		stateChangeEntry.Encoder = keyEncoder
+	}
+	// Set the encoder type.
+	stateChangeEntry.EncoderType = encoderType
+
+	// Encode the state change entry. We encode as a byte array, so the consumer can buffer just the bytes needed
+	// to decode this entry when reading from file.
+	entryBytes := EncodeToBytes(0, stateChangeEntry, false)
```

Comment:
hmmm we probably need a real block height in this call to `EncodeToBytes`, right? This gets passed down into `RawEncodeWithoutMetadata`